File: scripts/day10-part2-optimized.js
const fs = require("fs");

function parseMachine(line) {
  const match = line.match(/\[(.*?)\](.*)\{(.*)\}/);
  const lights = match[1];
  const buttonsStr = match[2].trim();
  const joltages = match[3].split(",").map(Number);

  const buttons = [];
  const buttonMatches = buttonsStr.matchAll(/\(([0-9,]+)\)/g);
  for (const m of buttonMatches) {
    buttons.push(m[1].split(",").map(Number));
  }

  return { lights, buttons, joltages };
}

function getAllParityMaps(nCounters, switches) {
  const parityMaps = new Map();

  for (let i = 0; i < 1 << switches.length; i++) {
    const result = new Array(nCounters).fill(0);
    let nFlips = 0;

    for (let j = 0; j < switches.length; j++) {
      if (i & (1 << j)) {
        nFlips++;
        for (const ele of switches[j]) {
          result[ele]++;
        }
      }
    }

    const parityKey = result.map((x) => x % 2).join(",");
    const resultKey = result.join(",");

    if (!parityMaps.has(parityKey)) {
      parityMaps.set(parityKey, new Map());
    }

    if (
      !parityMaps.get(parityKey).has(resultKey) ||
      parityMaps.get(parityKey).get(resultKey) > nFlips
    ) {
      parityMaps.get(parityKey).set(resultKey, nFlips);
    }
  }

  return parityMaps;
}

function findMinSwitchFlips(current, switches, parityMaps, cache) {
  const key = current.join(",");

  if (cache.has(key)) {
    return cache.get(key);
  }

  if (current.every((x) => x === 0)) {
    cache.set(key, { found: true, count: 0 });
    return { found: true, count: 0 };
  }

  if (current.some((x) => x < 0)) {
    cache.set(key, { found: false, count: 0 });
    return { found: false, count: 0 };
  }

  const currentParity = current.map((x) => x % 2).join(",");

  if (!parityMaps.has(currentParity)) {
    cache.set(key, { found: false, count: 0 });
    return { found: false, count: 0 };
  }

  let minFlips = Infinity;

  for (const [patternStr, nFlips] of parityMaps.get(currentParity)) {
    const pattern = patternStr.split(",").map(Number);

    // Check validity
    let valid = true;
    for (let i = 0; i < pattern.length; i++) {
      if (pattern[i] > current[i]) {
        valid = false;
        break;
      }
    }

    if (!valid) continue;

    const next = current.map((c, i) => (c - pattern[i]) / 2);
    const result = findMinSwitchFlips(next, switches, parityMaps, cache);

    if (result.found) {
      minFlips = Math.min(minFlips, nFlips + 2 * result.count);
    }
  }

  if (minFlips === Infinity) {
    cache.set(key, { found: false, count: 0 });
    return { found: false, count: 0 };
  }

  cache.set(key, { found: true, count: minFlips });
  return { found: true, count: minFlips };
}

function solvePart2(machines) {
  let total = 0;

  for (const machine of machines) {
    const parityMaps = getAllParityMaps(
      machine.joltages.length,
      machine.buttons
    );
    const cache = new Map();
    const result = findMinSwitchFlips(
      machine.joltages,
      machine.buttons,
      parityMaps,
      cache
    );
    total += result.count;
  }

  return total;
}

const input = fs
  .readFileSync("../inputs/day-10-input.txt", "utf-8")
  .trim()
  .split(/\r?\n/);

const machines = input.map(parseMachine);

const iterations = 1;
const startTime = process.hrtime.bigint();

let result;
for (let i = 0; i < iterations; i++) {
  result = solvePart2(machines);
}

const endTime = process.hrtime.bigint();
const totalNanos = Number(endTime - startTime);
const avgMicros = (totalNanos / iterations / 1000).toFixed(3);

console.log(`JavaScript (Day 10 Part 2 - Optimized)`);
console.log(`Answer: ${result}`);
console.log(`Average time: ${avgMicros} microseconds (${iterations} iterations)`);
console.log(`Total time: ${(totalNanos / 1000000).toFixed(2)} ms`);

File: scripts/day10-part2-optimized.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <regex>
#include <chrono>
#include <unordered_map>
#include <bitset>

using namespace std;

struct Machine {
    string lights;
    vector<vector<int>> buttons;
    vector<int> joltages;
};

struct VectorHasher {
    size_t operator()(const vector<int>& v) const {
        size_t hash = 0;
        for (int i : v) {
            hash = hash * 31 + i;
        }
        return hash;
    }
};

Machine parseMachine(const string& line) {
    Machine m;
    regex pattern(R"(\[(.*?)\](.*)\{(.*)\})");
    smatch match;
    regex_search(line, match, pattern);
    
    m.lights = match[1];
    string buttonsStr = match[2];
    string joltagesStr = match[3];
    
    size_t pos = 0;
    while ((pos = joltagesStr.find(',')) != string::npos) {
        m.joltages.push_back(stoi(joltagesStr.substr(0, pos)));
        joltagesStr.erase(0, pos + 1);
    }
    m.joltages.push_back(stoi(joltagesStr));
    
    regex buttonPattern(R"(\(([0-9,]+)\))");
    sregex_iterator iter(buttonsStr.begin(), buttonsStr.end(), buttonPattern);
    sregex_iterator end;
    
    while (iter != end) {
        vector<int> button;
        string btnStr = (*iter)[1];
        size_t p = 0;
        while ((p = btnStr.find(',')) != string::npos) {
            button.push_back(stoi(btnStr.substr(0, p)));
            btnStr.erase(0, p + 1);
        }
        button.push_back(stoi(btnStr));
        m.buttons.push_back(button);
        ++iter;
    }
    
    return m;
}

unordered_map<vector<int>, unordered_map<vector<int>, int, VectorHasher>, VectorHasher> 
getAllParityMaps(int nCounters, const vector<vector<int>>& switches) {
    unordered_map<vector<int>, unordered_map<vector<int>, int, VectorHasher>, VectorHasher> parityMaps;
    
    for (int i = 0; i < (1 << switches.size()); i++) {
        bitset<16> bits(i);
        vector<int> result(nCounters, 0);
        
        for (size_t j = 0; j < switches.size(); j++) {
            if (bits[j]) {
                for (int ele : switches[j]) {
                    result[ele]++;
                }
            }
        }
        
        vector<int> parityKey;
        for (int ele : result) {
            parityKey.push_back(ele % 2);
        }
        
        int nFlips = bits.count();
        
        if (parityMaps[parityKey].find(result) == parityMaps[parityKey].end() ||
            parityMaps[parityKey][result] > nFlips) {
            parityMaps[parityKey][result] = nFlips;
        }
    }
    
    return parityMaps;
}

pair<bool, int> findMinSwitchFlips(
    const vector<int>& current,
    const vector<vector<int>>& switches,
    const unordered_map<vector<int>, unordered_map<vector<int>, int, VectorHasher>, VectorHasher>& parityMaps,
    unordered_map<vector<int>, pair<bool, int>, VectorHasher>& cache
) {
    auto it = cache.find(current);
    if (it != cache.end()) {
        return it->second;
    }
    
    bool allZero = true;
    for (int x : current) {
        if (x != 0) {
            allZero = false;
            break;
        }
    }
    if (allZero) {
        cache[current] = {true, 0};
        return {true, 0};
    }
    
    for (int x : current) {
        if (x < 0) {
            cache[current] = {false, 0};
            return {false, 0};
        }
    }
    
    vector<int> currentParity;
    for (int c : current) {
        currentParity.push_back(c % 2);
    }
    
    auto pmIt = parityMaps.find(currentParity);
    if (pmIt == parityMaps.end()) {
        cache[current] = {false, 0};
        return {false, 0};
    }
    
    int minFlips = 999999;
    
    for (const auto& [pattern, nFlips] : pmIt->second) {
        bool valid = true;
        for (size_t i = 0; i < pattern.size(); i++) {
            if (pattern[i] > current[i]) {
                valid = false;
                break;
            }
        }
        
        if (!valid) continue;
        
        vector<int> next;
        next.reserve(current.size());
        for (size_t i = 0; i < current.size(); i++) {
            next.push_back((current[i] - pattern[i]) / 2);
        }
        
        auto [found, count] = findMinSwitchFlips(next, switches, parityMaps, cache);
        
        if (found) {
            minFlips = min(minFlips, nFlips + 2 * count);
        }
    }
    
    if (minFlips == 999999) {
        cache[current] = {false, 0};
        return {false, 0};
    }
    
    cache[current] = {true, minFlips};
    return {true, minFlips};
}

int solvePart2(const vector<Machine>& machines) {
    int total = 0;
    
    for (const auto& machine : machines) {
        auto parityMaps = getAllParityMaps(machine.joltages.size(), machine.buttons);
        unordered_map<vector<int>, pair<bool, int>, VectorHasher> cache;
        auto [found, count] = findMinSwitchFlips(machine.joltages, machine.buttons, parityMaps, cache);
        total += count;
    }
    
    return total;
}

int main() {
    ifstream file("../inputs/day-10-input.txt");
    vector<Machine> machines;
    machines.reserve(200);
    string line;
    
    while (getline(file, line)) {
        machines.push_back(parseMachine(line));
    }
    file.close();
    
    const int iterations = 1;
    auto start = chrono::high_resolution_clock::now();
    
    int result;
    for (int i = 0; i < iterations; i++) {
        result = solvePart2(machines);
    }
    
    auto end = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::nanoseconds>(end - start);
    double avgMicros = duration.count() / (double)iterations / 1000.0;
    
    cout << "C++ (Day 10 Part 2 - Optimized)" << endl;
    cout << "Answer: " << result << endl;
    cout << "Average time: " << avgMicros << " microseconds (" << iterations << " iterations)" << endl;
    cout << "Total time: " << (duration.count() / 1000000.0) << " ms" << endl;
    
    return 0;
}
